### PR TITLE
Fixed wrong formatting of docker logs when being collected on Kubernetes from /var/log/pods

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

Make sure log line from docker container files are properly parsed when being collected.

### Motivation

Make sure logs show up properly on the logs explorer.

### Additional Notes

If the log exceeds 16kb, it will be truncated, this has to be fixed for `6.13`

This should fix https://github.com/DataDog/datadog-agent/issues/3510

### CODE REVIEW

File: pkg/logs/input/docker/json_parser.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// returns an error if it failed.
+// For example:
+// {"log":"a message","stream":"stderr","time":"2019-06-06T16:35:55.930852911Z"}
+// returns:
+// "a message", "error", "2019-06-06T16:35:55.930852911Z", nil
+func (p *jsonParser) Parse(data []byte) ([]byte, string, string, error) {
+	var log *logLine
+	err := json.Unmarshal(data, &log)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("could not parse JSON container line: %v", err)
```
Comment (achntrl) [inline pkg/logs/input/docker/json_parser.go]: Should we return the raw line if we failed to parse it ?

```suggestion
		return data, "", "", fmt.Errorf("could not parse JSON container line: %v", err)
```

**Review by achntrl — COMMENTED** [review]
What happens when the line is more than 16kb ? is the JSON broken and we don't receive a valid JSON ?

File: pkg/logs/input/docker/json_parser.go (hunk 1 above)
Comment (ajacquemot) [reply pkg/logs/input/docker/json_parser.go]: 👍

**Review by achntrl — DISMISSED** [review]

**Review by achntrl — DISMISSED** [review]

File: pkg/logs/input/docker/json_parser.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// returns an error if it failed.
+// For example:
+// {"log":"a message","stream":"stderr","time":"2019-06-06T16:35:55.930852911Z"}
+// returns:
+// "a message", "error", "2019-06-06T16:35:55.930852911Z", nil
+func (p *jsonParser) Parse(data []byte) ([]byte, string, string, error) {
+	var log *logLine
+	err := json.Unmarshal(data, &log)
+	if err != nil {
+		return data, message.StatusInfo, "", fmt.Errorf("could not parse JSON container line: %v", err)
```
Comment (tt810) [inline pkg/logs/input/docker/json_parser.go]: What is `JSON container line`? could you rephrase it?

File: pkg/logs/input/docker/json_parser_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
+)
+
+func TestJSONParser(t *testing.T) {
```
Comment (tt810) [inline pkg/logs/input/docker/json_parser_test.go]: could you test the case when 
* a field is empty (if `log` field is empty, is that a valid case?)
* a field absent

File: releasenotes/notes/logs-container-configuration-toggle-f5f8b2831c6a575f.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+Breaking change - In the version 6.11.2 logic was added in the Agent to first look for K8s container files if `/var/log/pods` was not available and then to go for the Docker socket.
+This created some permission issues as `/var/log/pods` can be a symlink in some configuration and the Agent also needed access to the symlink directory.
+
+This logic is reverted to its prior behaviour which prioritise the Docker socket for container log collection.
+It is still possible to force the agent to go for the K8s log files even if the Docker socket is mounted by using the `logs_config.k8s_container_use_file' or `DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE`. parameter.
+This is recommend when more than 10 containers are used on the same pod.
```
Comment (tt810) [inline releasenotes/notes/logs-container-configuration-toggle-f5f8b2831c6a575f.yaml]:
```suggestion
This is recommend when more than 10 containers are running in the same pod.
```

File: pkg/logs/input/file/tailer.go [hunk 5]
```diff
@@ -58,9 +59,12 @@ type Tailer struct {
 func NewTailer(outputChan chan *message.Message, source *config.LogSource, path string, sleepDuration time.Duration, isWildcardPath bool) *Tailer {
 	// TODO: remove those checks and add to source a reference to a tagProvider and a lineParser.
 	var parser lineParser.Parser
-	if source.GetSourceType() == config.KubernetesSourceType {
+	switch source.GetSourceType() {
+	case config.KubernetesSourceType:
 		parser = kubernetes.Parser
-	} else {
+	case config.DockerSourceType:
```
Comment (tt810) [inline pkg/logs/input/file/tailer.go]: Could you add some tests related to the issue we discussed earlier: 
* logs splited by docker when oversized. so 2 independent logs are passed to decoder the later one without '\n' in the message. 
* decoder with multilinehandler to see the behaviour of this parser.

**Review by tt810 — CHANGES_REQUESTED** [review]
More tests need to be added.

File: pkg/logs/input/docker/json_parser_test.go (hunk 3 above)
Comment (ajacquemot) [reply pkg/logs/input/docker/json_parser_test.go]: no need to test when a field is absent because by default, when a field is absent, go would put a null `value`, same for `unrecognised` field, by design it won't fail.
If the log is empty, the parsing will succeed as well.

File: pkg/logs/input/file/tailer.go (hunk 5 above)
Comment (tt810) [reply pkg/logs/input/file/tailer.go]: I don't see the tests for the 1st scenario

File: pkg/logs/decoder/line_handler.go [hunk 6]
```diff
@@ -194,18 +195,19 @@ func (h *MultiLineHandler) run() {
...
-	h.lastSeenTimestamp = timestamp
+	unwrappedLine, status, timestamp, err := h.parser.Parse(line)
 	if err != nil {
 		log.Debug(err)
 	}
 	if h.newContentRe.Match(unwrappedLine) {
 		// send content from lineBuffer
 		h.sendContent()
 	}
+	h.lastSeenTimestamp = timestamp
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/decoder/line_handler.go]: we should fix this later on

File: pkg/logs/decoder/decoder_test.go [hunk 7]
```diff
@@ -170,3 +172,47 @@ func TestDecoderWithDockerHeader(t *testing.T) {
...
+func TestDecoderWithSingleline(t *testing.T) {
+	var output *Output
+
+	d := InitializeDecoder(config.NewLogSource("", &config.LogsConfig{}), parser.NoopParser)
+	d.Start()
+	defer d.Stop()
+
+	d.InputChan <- NewInput([]byte("message\n"))
+	output = <-d.OutputChan
+	assert.Equal(t, []byte("message"), output.Content)
```
Comment (tt810) [inline pkg/logs/decoder/decoder_test.go]: Why "\n" is removed here? where is this removal logic?

File: pkg/logs/input/docker/json_parser.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		status = message.StatusError
+	case stdout:
+		status = message.StatusInfo
+	default:
+		status = ""
+	}
+
+	content := []byte(log.Log)
+	length := len(content)
+	if length > 0 && log.Log[length-1] == '\n' {
```
Comment (tt810) [inline pkg/logs/input/docker/json_parser.go]: Remind me the reason we are doing this check again?

Comment (ajacquemot) [reply pkg/logs/input/docker/json_parser.go]: to remove the ending `\n` as we don't want to keep it for now, otherwise it would add an extra empty line for multiline logs.

File: pkg/logs/decoder/decoder_test.go (hunk 7 above)
Comment (ajacquemot) [reply pkg/logs/decoder/decoder_test.go]: https://github.com/DataDog/datadog-agent/pull/3655#discussion_r291734553

File: releasenotes/notes/logs-docker-kubernetes-format-fix-8f0101796f2fe46c.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Each section from every releasenote are combined when the
```
Comment (tt810) [inline releasenotes/notes/logs-docker-kubernetes-format-fix-8f0101796f2fe46c.yaml]: should we remove this comment?

**Review by tt810 — COMMENTED** [review]
LGTM. small comment in the yaml file.

File: releasenotes/notes/logs-docker-kubernetes-format-fix-8f0101796f2fe46c.yaml (hunk 9 above)
Comment (ajacquemot) [reply releasenotes/notes/logs-docker-kubernetes-format-fix-8f0101796f2fe46c.yaml]: it's removed automatically when re-aggregating release notes, no need

**Review by tt810 — APPROVED** [review]
